src/kixi/spec.clj

```diff
@@ -39,3 +42,76 @@
        (catch Exception _
          (create-ns namespace-sym)
          (clojure.core/alias alias namespace-sym))))
+
+(alias 'c 'clojure.core)
+
+(defn- unfn [expr]
+  (if (c/and (seq? expr)
+             (symbol? (first expr))
+             (= "fn*" (name (first expr))))
+    (let [[[s] & form] (rest expr)]
+      (conj (walk/postwalk-replace {s '%} form) '[%] 'fn))
+    expr))
+
+(defn- ->sym
+  "Returns a symbol from a symbol or var"
+  [x]
+  (if (var? x)
+    (let [^clojure.lang.Var v x]
+      (symbol (str (.name (.ns v)))
+              (str (.sym v))))
+    x))
+
+(defn- res [form]
+  (cond
+    (keyword? form) form
+    (symbol? form) (c/or (-> form resolve ->sym) form)
+    (sequential? form) (walk/postwalk #(if (symbol? %) (res %) %) (unfn form))
+    :else form))
+
+(defn all-preds-spec-impl
+  [keys forms preds gfn]
+  (let [id (java.util.UUID/randomUUID)
+        kps (zipmap keys preds)
+        cform (fn [x]
+                (loop [[k p] (first kps)]
+                  (if (and k p)
+                    (let [ret (p x)]
+                      (if ret
+                        (recur (rest kps))
+                        ::s/invalid))
```

Comment:
Without the tagged ret value how does the error get propagated?

https://github.com/clojure/clojure/blob/d920ada9fab7e9b8342d28d8295a600a814c1d8a/src/clj/clojure/spec.clj#L925